### PR TITLE
Translate botnet to Rust

### PR DESCRIPTION
**Pull request by equal-l2**
I've translated botnet module into Rust.
As the translation is almost word-to-word, please feel free to tell me if I can rewrite something in more rust-ish way.

### CODE REVIEW

File: src/botnet.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    for (i, nodes) in clusters.iter().enumerate() {
+        dprint(format!("  Cluster #{:02} ({:3} nodes)\n", i, nodes), 10);
+        csleep(100);
+    }
+
+    loop {
+        print!("\u{001b}[{}A\n", onlines.len() + 1);
+        for (i, (nodes, online)) in clusters.iter().zip(onlines.iter()).enumerate() {
+            // cannot use bright color because yansi doesn't support it
```
Comment (svenstaro) [inline src/botnet.rs]: I think it does in the form of `.bold()`. Can you check?

File: src/botnet.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        for o in &mut onlines {
+            if rng.gen_range(0., 1.) > 0.95_f64 {
+                *o = true;
+            }
+        }
+        csleep(100);
+    }
+
+    for task in &[
```
Comment (svenstaro) [inline src/botnet.rs]: Maybe break this out into a separate variable so the loop looks nicer?

File: src/botnet.rs (hunk 1 above)
Comment (equal-l2) [reply src/botnet.rs]: It does indeed.
I'll fix now.

**Comment by equal-l2** [comment]
Please tell me if I should squash all commits before merging.

**Comment by svenstaro** [comment]
Really cool stuff! I had a proper look. Review to follow.

**Comment by equal-l2** [comment]
Thank you for detailed advice.
I'll look into and fix later.

**Comment by equal-l2** [comment]
Actually I'm poor at giving names, so please tell me if you have more appropriate names.

**Comment by svenstaro** [comment]
Good changes! However, it still looks broken in the web version. Can you have a look? All you have to type is `make run-web` and then connect to `localhost:8000` with a wasm-enabled browser.

**Comment by svenstaro** [comment]
screenshot-localhost-8000 2018 06 05 20-46-53

**Comment by svenstaro** [comment]
For your information, genact uses xterm.js as a virtual xterm-compatible console in the web so everything that can be done with ANSI escape codes should also be possible there. Writing happens in `utils.rs`.

File: src/botnet.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            10,
+        );
+        csleep(100);
+        if appconfig.should_exit() {
+            return;
+        }
+    }
+
+    loop {
+        dprint(format!("\u{001b}[{line_up:}A", line_up = onlines.len()), 0);
```
Comment (svenstaro) [inline src/botnet.rs]: Wouldn't it be nice if you made some cursor movement utilities instead and used them in your code? Those would be nice additions to `utils.rs` and would make this a lot more readable.

**Comment by svenstaro** [comment]
After a short investigation I'm pretty sure things would just work if you used ANSI control codes instead of unicode ones.

**Comment by svenstaro** [comment]
I checked out your newest changes but it still appears to be having the cursor movement problems.

**Comment by equal-l2** [comment]
Hmm, control codes aren't the cause then.
I'll look into more deeper.

**Comment by svenstaro** [comment]
Almost there. Now it looks like this: 
screenshot-localhost-8000-2018 06 09-04-45-59

**Comment by equal-l2** [comment]
That's odd, it works fine when I run on my Firefox Nightly or Chrome Canary.

**Comment by svenstaro** [comment]
I'm sorry, my build dir was dirty. It does look great now! Thanks a lot for the PR.

**Pull request merged**